sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeFormatterHelper.scala

```diff
@@ -31,17 +31,50 @@ import org.apache.spark.sql.internal.SQLConf
 import org.apache.spark.sql.internal.SQLConf.LegacyBehaviorPolicy._
 
 trait DateTimeFormatterHelper {
+  private def getOrDefault(accessor: TemporalAccessor, field: ChronoField, default: Int): Int = {
+    if (accessor.isSupported(field)) {
+      accessor.get(field)
+    } else {
+      default
+    }
+  }
+
+  protected def toLocalDate(accessor: TemporalAccessor, allowMissingYear: Boolean): LocalDate = {
+    val year = if (accessor.isSupported(ChronoField.YEAR)) {
+      accessor.get(ChronoField.YEAR)
+    } else if (allowMissingYear) {
+      // To keep backward compatibility with Spark 2.x, we pick 1970 as the default value of year.
+      1970
+    } else {
+      throw new SparkUpgradeException("3.0",
+        "Year must be given in the date/timestamp string to be parsed. You can set " +
+          SQLConf.LEGACY_ALLOW_MISSING_YEAR_DURING_PARSING.key + " to true, to pick 1970 as " +
+          "the default value of year.", null)
+    }
+    val month = getOrDefault(accessor, ChronoField.MONTH_OF_YEAR, 1)
```

Comment:
Are we also going to error out if they specify the day but not the month? Really, the only formats that make sense are the ones where a full prefix is given in the y-m-d h-m-s sequence, and all others are likely to be a case where they made a mistake (e.g. asked for "mm" twice where they meant MM).

Reply:
January should be fine? It accepts any DAY_OF_MONTH.

Reply:
> asked for "mm" twice where they meant MM

This will error out unless the 2 mm parse to the same value, e.g. `mm mm yyyy` with value `10 10 1990`.

Reply:
I'm saying it works (i.e., it accepts all valid values) but it's nonsensical; it'll never return the complete timestamp that people intended.

That said, I can see cases where people want to parse contiguous ranges (e.g. just a time, or just a month, or just a month + day) and then extract the parsed components using `EXTRACT` / `date_part`, or just to format them differently.

So I guess the most permissive but still safe thing here would be to forbid missing "year" only if someone also requests "day" and "month", because that's the situation where "year" affects the valid range of "day". But if they don't request year *and* they don't request month, then month would default to 1 and there would be no impact on the valid range of days, it would always go up to 32. I.e.:

`"yyyy MM dd"` -> OK
`"MM dd"` -> fail, need year to know valid range of `dd`
`"MM"` -> OK
`"dd"` -> OK
`"hh:mm:ss"` -> OK
`"MM dd hh:mm:ss"` -> fail, need year to know valid range of `dd`

Reply:
ah this looks better!

I'm wondering if we still need the legacy config though. It's very hard to hit the missing year exception, and if it happens it's very likely to be a data error. Users should fix the data or change the pattern and provide the default year.

Reply:
I guess we could even allow the default year, and if people hit Feb 29 then they'll get an error at runtime. No need for a legacy config in that case...

Reply:
ok so we should go back to the first version, which just uses 1970 as the default value and fails for invalid dates at runtime 😂

Reply:
BTW the current framework just returns null if the parser fails. Shall we treat this case specially and throw an error that fails the query?

Reply:
Error out in ANSI mode only, I'd say, just like with all other parse failures (which should fail in ANSI mode -- I hope they already do).

FWIW, going back to the first version is not a shame now that we know it's the right thing to do. :) Maybe we can leave some comments in the code to explain the reasoning why we don't error out if things are missing.

Reply:
unfortunately datetime functions don't respect the ANSI flag at all. Maybe we should keep everything the same as Spark 2.4 (return null when parsing Feb 29 because 1970-02-29 is invalid), and implement ANSI behavior in Spark 3.1?

Reply:
Yikes! It seems terrible that we have to do that later, because then it can break existing workloads. Launching the ANSI behavior was our chance to do the right thing here. If we can still fix this in 3.0 then that would of course be preferable, but that seems like a stretch...